### PR TITLE
Fix SROA miscompile in large functions

### LINKED ISSUE #46408: Segfault/assertion failure/IR verification error due to non-dominated use of SSAValue
As seen on PkgEval:
- https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_date/2022-08/17/Gnuplot.primary.log
- https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_date/2022-08/17/StringAnalysis.primary.log

```
[15] signal (11): Segmentation fault
in expression starting at /home/pkgeval/.julia/packages/Gnuplot/Iz8LH/test/runtests.jl:72
jl_iintrinsic_2 at /cache/build/default-amdci5-1/julialang/julia-master/src/runtime_intrinsics.c:985
<= at ./int.jl:488 [inlined]
in at ./range.jl:1413 [inlined]
map at ./tuple.jl:299 [inlined]
iterate at ./multidimensional.jl:396 [inlined]
arrays2datablock at /home/pkgeval/.julia/packages/Gnuplot/Iz8LH/src/Gnuplot.jl:430
```

Looking at the `rr` trace, turns out we're passing garbage into `sle_int`:

```
Thread 1 received signal SIGSEGV, Segmentation fault.
0x00007fee329f6c71 in jl_iintrinsic_2 (a=0x7fee2c5f2060, b=0x3, name=0x7fee32b92e15 "sle_int", lambda2=0x7fee329f5860 <jl_intrinsiclambda_cmp>, list=0x7fee32c412a0 <sle_int_list>, cvtb=0, getsign=0x7fee329f5840 <signbitbyte>) at /cache/build/default-amdci5-1/julialang/julia-master/src/runtime_intrinsics.c:985
985	    jl_value_t *tyb = jl_typeof(b);
(rr) p b
$7 = (jl_value_t *) 0x3
```

```llvm
L3208:                                            ; preds = %L3207
  %3973 = call {} addrspace(10)* @jl_sle_int({} addrspace(10)* addrspacecast ({}* inttoptr (i64 140660923375712 to {}*) to {} addrspace(10)*), {} addrspace(10)* undef), !dbg !3060
  %3974 = bitcast {} addrspace(10)* %3973 to i8 addrspace(10)*, !dbg !3063
  %3975 = load i8, i8 addrspace(10)* %3974, align 1, !dbg !3063, !tbaa !3064, !range !3065
  %3976 = trunc i8 %3975 to i1, !dbg !3063
  %3977 = and i1 true, %3976, !dbg !3063
  %3978 = call {} addrspace(10)* @jl_sle_int({} addrspace(10)* addrspacecast ({}* inttoptr (i64 140660923375712 to {}*) to {} addrspace(10)*), {} addrspace(10)* undef), !dbg !3060
  %3979 = bitcast {} addrspace(10)* %3978 to i8 addrspace(10)*, !dbg !3063
  %3980 = load i8, i8 addrspace(10)* %3979, align 1, !dbg !3063, !tbaa !3064, !range !3065
  %3981 = trunc i8 %3980 to i1, !dbg !3063
  %3982 = and i1 true, %3981, !dbg !3063
  %3983 = and i1 %3977, %3982, !dbg !3066
  %3984 = xor i1 %3983, true, !dbg !3059
  %3985 = zext i1 %3984 to i8, !dbg !3057
  %3986 = trunc i8 %3985 to i1, !dbg !3057
  %3987 = xor i1 %3986, true, !dbg !3057
  br i1 %3987, label %L3216, label %L3215, !dbg !3057
```

Note the `undef` arguments.

Running the above under an assertions build reveals a failed assertion that can be reduced to:

```julia
using InteractiveUtils

function e(a::Vararg)
    b = 0
    for c in a
        0 && (b = c)
        if for d in CartesianIndices(a[b]) end
            for d in CartesianIndices(size(b)) end
        end
    end
end

code_llvm(e, Tuple{Array{Int64, 3}, Array{Int64, 2}})
```

```
julia: /home/tim/Julia/src/julia/src/codegen.cpp:4573: void emit_ssaval_assign(jl_codectx_t&, ssize_t, jl_value_t*): Assertion `!ctx.ssavalue_assigned.at(ssaidx_0based)' failed.
```

Bisected to #44557, so cc @ianatol.

### PR DESCRIPTION
**Pull request by Keno**
During the development of #46775, we saw a miscompile in the compiler, specifically, we saw SROA failing to provide a proper PHI nest. The root cause of this turned out to be an incorrect dominance query. In particular, during incremental compaction, the non-compacted portion of the IncrementalCompact cfg is allocated, but not valid, so we must not query it for basic block information.

Fixes #46408

### CODE REVIEW

**Comment by Keno** [comment]
> Fixes #46408

I'm pretty sure at least - the trace in that issue looks basically identical to what I was seeing. Unfortunately, the test case no longer seems to reproduce the issue on master (which I'm unsurprised by - this error is very sensitive to the exact order of things).

**Comment by Keno** [comment]
@maleadt was there an unreduced version of this that might result in a still-broken reproducer?

File: base/compiler/ssair/ir.jl [hunk 1]
```diff
@@ -28,12 +28,18 @@ function cfg_delete_edge!(cfg::CFG, from::Int, to::Int)
     nothing
 end
 
+function bb_ordering()
+    lt=(<=)
+    by=x->first(x.stmts)
+    ord(lt, by, nothing, Forward)
+end
```
Comment (aviatesk) [inline base/compiler/ssair/ir.jl:35]: How about making this constant?
```suggestion
const BB_ORDERING = ord((<=), x->first(x.stmts), nothing, Forward)
```

**Comment by aviatesk** [comment]
Thanks so much for fixing this. It looks like this PR fixes the problem we saw on https://github.com/JuliaLang/julia/pull/46775.

**Comment by aviatesk** [comment]
@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/46819#issuecomment-1250238930) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/3a82b41_vs_f794bdd/report.md).

**Comment by maleadt** [comment]
> @maleadt was there an unreduced version of this that might result in a still-broken reproducer?

I didn't keep any intermediaries, but IIRC I started from the `arrays2datablock` function in Gnuplot.jl (which is called a bunch from their test suite).

**Comment by ianatol** [comment]
Seems reasonable, thanks for getting to this Keno! I'd still like to recreate the dominance error just to make sure this actually fixes it, but it seems like it should.

**Comment by KristofferC** [comment]
This needs a manual backport to 1.8.3

**Pull request merged**